File: src/2017/day03/python/solution.py
def format_data(input: str) -> list[tuple[int, ...]]:
    lines = [tuple(map(int, line.split())) for line in input.strip().splitlines()]
    return lines


def part1(input: str) -> int:
    return sum(max(nums) - min(nums) for nums in format_data(input))


def get_even_divisor(nums: tuple[int, ...]) -> int:
    sorted_nums = sorted(nums, reverse=True)

    generator = (
        a // b
        for a_index, a in enumerate(sorted_nums)
        for b in sorted_nums[a_index + 1 :]
        if a % b == 0
    )

    return next(generator, 0)


def part2(input: str) -> int:
    return sum(get_even_divisor(line) for line in format_data(input))

File: src/2017/day03/typescript/solution.ts
function formatData(input: string): number[][] {
  return input
    .trim()
    .split('\n')
    .map((line) => line.trim().split(/\s/g).map(Number));
}

export function part1(input: string): number | string | undefined {
  return formatData(input)
    .map((line) => Math.max(...line) - Math.min(...line))
    .reduce((acc, num) => acc + num, 0);
}

function getEvenDivisor(nums: number[]): number {
  nums.sort((a, b) => b - a);
  const N = nums.length;

  for (let l = 0; l < N; l++) {
    for (let r = l + 1; r < N; r++) {
      const numLeft = nums[l];
      const numRight = nums[r];

      if (numLeft % numRight === 0) {
        return numLeft / numRight;
      }
    }
  }
}

export function part2(input: string): number | string | undefined {
  return formatData(input)
    .map((line) => getEvenDivisor(line))
    .reduce((acc, num) => acc + num);
}
